WIP - introduce ratcheting validation mechanism

liggitt opened the pull request:
**What this PR does / why we need it**:

Introduces a ratcheting validation mechanism, as described in https://github.com/kubernetes/kubernetes/issues/64841#issuecomment-395141509

Tightening validation on existing data cannot be done in a way that prevents existing stored objects from being updated.

This PR introduces the following mechanism:
* identify granular additional validations (usually to fix fatal errors that always result in persisted, yet broken objects)
* apply those validations on create to prevent new broken objects from being created
* apply those validations on update **only if the existing object also passes the validation**
  * this prevents valid objects from becoming invalid
  * this avoids breaking the ability to update existing invalid objects

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64841 
Fixes #57510 - flex driver name
Fixes #58477 - duplicate envvar names
Fixes #54567, #64011 - invalid memory quantities (xref #63426)
xref #52936 - storage medium
xref #60934 - duplicate pvc claimName

**Special notes for your reviewer**:
/assign @lavalamp @deads2k 

**Release note**:
```release-note
Validation has tightened when creating new pods (or new objects containing pod templates):
* flex volume driver names must now be valid
* emptyDir medium must now be valid
* duplicate envvar names are not allowed
* duplicate volumes referencing the same PVC claimName are not allowed
* memory limits and requests must be an integer value
```

liggitt committed:
Add ratcheting validation mechanism

validate flex volume driver name

validate emptyDir medium

prevent duplicate envvar names

validate duplicate PVC volume names

validate memory units are not fractional

liggitt commented:
@kubernetes/sig-api-machinery-pr-reviews

liggitt commented:
would like feedback on the mechanism first, then will tag various teams on the specific validation commits (those were mostly to illustrate how the mechanism could be used)

liggitt commented:
/retest

lavalamp commented on pkg/apis/core/validation/ratcheting.go:
```diff
@@ -60,4 +60,31 @@ var RatchetingPodSpecValidations = apimachineryvalidation.RatchetingValidators{
+		spec := data.(*core.PodSpec)
+		for i, c := range spec.InitContainers {
+			names := sets.NewString()
+			for j, e := range c.Env {
```
nit: seems like these loops could be a subroutine.

liggitt replied:
cleaned up the loops

lavalamp commented:
Looks good overall.

I do wish there were less setup code: folks have to remember to call this everywhere there's a pod template type. But it is fairly straightforward, maybe that is better than a complicated automated system. We can always automate when we find people making errors that the automation would catch.

sttts commented:
> We should think about how to do this problem for CRD validation, too, when we add it. Perhaps there all validations should be ratcheting.

This is getting interesting. For native types we only have positive validation predicates `A(old,new) and B(old,new) and C(old,new)`. So we can easily declare some as ratcheting (call it `A'`, `B'`, `C'`), making the complete validation predicate `A'(old,new) and B'(old,new) and C'(old,new)` weaker than the original. With CRDs and their OpenAPI based validation, we don't have positivity, e.g. `((not A(old,new)) or B(old,new)) and C(old,new)` is expressible. Making `A` ratcheting as `A'` gives a stronger predicate than before, not what we want. So ratcheting has to be restricted to positive predicates.

Another speciality for CRDs: for defaulting we have to apply the OpenAPI schema before we even have an old object to check against. We could get away by tagging certain OpenAPI properties as ratcheting, drop them on (coercing) schema validation for defaulting and only apply them on strategy validation. 

But even this way, it is not obvious how a ratcheting validation with an `old` and a `new` object will look like. The validation algorithm and the whole OpenAPI schema semantics does not know the concept of validation of differences between two objects on update. If we come up with such semantics though, this would also be very interesting to formulate "read-only field", which cannot be espressed today in OpenAPI at all.

lavalamp commented:
@sttts I assumed we would be writing a validator that takes both old and
new objects, and only enforces a validation check on the new object if the
old one passes (or is missing the relevant field). I am talking about value
validation and not schema validation, of course.

And yes, we will definitely be implementing read-only/immutable fields, for
built-ins and CRs. Take a look at some of the stuff @apelisse and @seans3
have been doing in kube-openapi if you haven't seen it already.

On Mon, Jun 11, 2018 at 1:35 AM Dr. Stefan Schimanski <
notifications@github.com> wrote:

> We should think about how to do this problem for CRD validation, too, when
> we add it. Perhaps there all validations should be ratcheting.
>
> This is getting interesting. For native types we only have positive
> validation predicates A(old,new) and B(old,new) and C(old,new). So we can
> easily declare some as ratcheting (call it A', B', C'), making the
> complete validation predicate A'(old,new) and B'(old,new) and C'(old,new)
> weaker than the original. With CRDs and their OpenAPI based validation, we
> don't have positivity, e.g. ((not A(old,new)) or B(old,new)) and
> C(old,new) is expressible. Making A ratcheting as A' gives a stronger
> predicate than before, not what we want. So ratcheting has to be restricted
> to positive predicates.
>
> Another speciality for CRDs: for defaulting we have to apply the OpenAPI
> schema before we even have an old object to check against. We could get
> away by tagging certain OpenAPI properties as ratcheting, drop them on
> (coercing) schema validation for defaulting and only apply them on strategy
> validation.
>
> But even this way, it is not obvious how a ratcheting validation with an
> old and a new object will look like. The validation algorithm and the
> whole OpenAPI schema semantics does not know the concept of validation of
> differences between two objects on update. If we come up with such
> semantics though, this would also be very interesting to formulate
> "read-only field", which cannot be espressed today in OpenAPI at all.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/64907#issuecomment-396164938>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngloo5xoZ7tDp7wJ4Tuc7jh4WNh3uLks5t7ivlgaJpZM4Uflyh>
> .
>

liggitt commented:
> @sttts I assumed we would be writing a validator that takes both old and new objects, and only enforces a validation check on the new object if the old one passes (or is missing the relevant field). I am talking about value validation and not schema validation, of course.

One of the reasons the approach in this PR works well is that the tightening validations are separable. Take the following scenario:

* I have base validation V, and tightening validations A, B, C.
* I have an old object that passes V, A, B, and fails C

Any update would also be required to pass V, A, B. If V, A, B, and C could not be checked individually, the old object failing C means we'd have to skip all validation on the updated object, which would open us up to objects with much bigger validation problems.

@sttts is the declarative validation able to be broken down to the level of individual checks?

mbohlool commented:
/sub

sttts commented:
> @sttts is the declarative validation able to be broken down to the level of individual checks?

No, in general it is not. But one knows which parts of a schema can be "broken apart" in the way you describe. So I think we can mark certain sub-schemata as ratcheting and the CRD type validation will check those markers. E.g. in `((not A(old,new)) or B(old,new)) and C(old,new)` only `B` and `C` are positive and therefore can be ratcheting (or some sub-schema of them).

sttts commented:
> And yes, we will definitely be implementing read-only/immutable fields, for
> built-ins and CRs. Take a look at some of the stuff @apelisse and @seans3
> have been doing in kube-openapi if you haven't seen it already.

@lavalamp If you have a link, that would be awesome. This sounds interesting. I saw the extension PRs. Do you mean those to express new kube-specific properties?

We can certainly make this work using the restriction to positive sub-schemata (see post above). I am not completely convinced though yet that this always gives the semantics we want and that users expect, especially because schema and value validation are not separable: some value sub-schemata are required to select the right branches of `AllOf/AnyOf/OneOf` quantors. This needs some more thought.

yliaog commented:
/cc @yliaog

fedebongio commented:
/cc @cheftako @jpbetz

lavalamp commented:
> > @sttts is the declarative validation able to be broken down to the level of individual checks?

> No, in general it is not.

But that is just because no one has done it. It's not much harder to walk two objects than it is to walk one. OpenAPI validation rules are all on single fields (right?) so in principle they are totally separable.

sttts commented:
> But that is just because no one has done it. It's not much harder to walk two objects than it is to walk one.

About the algorithm I agree. The old+new validation walker just has to be done in go-openapi/validate, and it's not much harder.

What I meant is the semantics:

- assume `{allOf {not: {pattern: "a"}, {type: "string"}}}` as a schema, e.g. `not: {pattern: "a"}` was added as a new validation. This validates `"b"`, because `"b"` matches the second clause (type), but not the first (pattern).
- this corresponds to the predicate: `!patternMatches(new, "a") && isString(new)` – as propositional formula `~A and B`
- add general ratcheting to each atom of the predicate: `!(!patternMatches(old, "a") || patternMatches(new, "a")) && (!isString(old) || isString(new))` – as propositional formula `~(A' => A) and (B' => B)`.
- this is equivalent to `!!(!!patternMatches(old, "a") && !patternMatches(new, "a")) && (!isString(old) || isString(new))` – as propositional formula `~~(~~A' and ~A) and (B' => B)`.
- and can be simplified to `patternMatches(old, "a") && !patternMatches(new, "a") && (!isString(old) || isString(new))` – as propositional formula `A' and ~A and (B' => B)`.

Notice that general ratcheting on all atoms made this formula stricter, not weaker. Take `new := "b", old := "b"` as an example again:
- without ratcheting `!patternMatches(new, "a") && isString(new)` gives `true`
- with ratcheting `patternMatches(old, "a") && !patternMatches(new, "a") && (!isString(old) || isString(new))` gives `false`. 

I.e. `"b"` – which was totally valid before – is now invalid. It's not what we want. In other words, we cannot apply ratcheting blindly to negative sub-schematas in a schema.

sttts commented:
Thinking out loud:

We could split value and schema validation within one schema like:
```
{ properties: { "name": { type: "string", pattern: "[a-z]+", "x-value": [{anyOf: [{pattern: "foo"}, {pattern: "bar"}], not: {...}}]}}}
```
I.e. the schema pattern part is a "simple" schema, only positive sub-schemata, no quantors. Under `x-value` the schema is free. Ratcheting is applied to each `x-value` sub-schema, but not pulled into its atoms.

A split of schema and value validation would also enforce that schema validation complexity is not exploding (we call this often in our decoding/conversion stack, but value validation just once).

lavalamp commented:
@sttts 

hot take: Yikes, you can mix value and schema validation??!?!?! That sounds crazy at first. Is this for saying things like "this should be a string-formatted IP or a 32 bit number packed IP"? Are any such things reasonable to say? We have IntOrString, which is also crazy, but even we are not crazy enough to have value and schema requirements at the same time. Mixing ontological levels up in the type system seems like a realllllly bad idea.

Let's assume we don't allow that. In our system, schema validation is about the *type* of the object, and the object has to validate to be interpretable. Then things like admission control run, and finally value validation, which includes verifying that required fields are present.

Given that (which I think is both how our system should and does work), can you still construct your example?

nikhita commented:
Is there anything blocking this PR apart from the discussion for CRD validation (which I assume can be handled separately)?

nikhita commented:
Would love to use it for https://github.com/kubernetes/kubernetes/pull/65472#discussion_r198615252.

liggitt commented:
Nothing blocking it, just getting back to it after 1.11 cuts

liggitt commented:
rebased and updated. tagging in sigs for review of area-specific validation tightening

@kubernetes/sig-storage-pr-reviews:
* `validate flex volume driver name`
* `validate emptyDir medium`
* `validate duplicate PVC volume names`

@kubernetes/sig-node-pr-reviews:
* `validate memory units are not fractional`

@kubernetes/sig-api-machinery-pr-reviews:
* `prevent duplicate envvar names` (broke apply/update removals)

liggitt commented:
cc @kubernetes/api-reviewers @kubernetes/api-approvers 

this gives us a way to prevent new objects that are fatally flawed, while not causing existing persisted data to be rejected during update/delete requests required to migrate storage and clean them up

sttts commented on staging/src/k8s.io/apimachinery/pkg/api/validation/ratcheting.go:
```diff
@@ -0,0 +1,44 @@
+	return allErrs
+}
+
+func ValidateRatchetingUpdate(newData, existingData interface{}, fldPath *field.Path, validators RatchetingValidators) field.ErrorList {
```
go doc.

sttts commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -28,8 +28,9 @@ import (
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	apimachineryvalidation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
+	apimachineryv1validation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
```
nit: metav1validation

sttts commented on pkg/apis/core/validation/ratcheting.go:
```diff
@@ -0,0 +1,124 @@
+		names := sets.NewString()
+		for i, v := range spec.Volumes {
+			if v.VolumeSource.PersistentVolumeClaim != nil {
+				if names.Has(v.VolumeSource.PersistentVolumeClaim.ClaimName) {
```
nit: if cn := v.VolumeSource.PersistentVolumeClaim; cn != nil {

then use cn below. Easier to read.

sttts commented:
Looks good overall. Some nits. Please squash.

liggitt commented:
> Looks good overall. Some nits. Please squash.

Fixed nits. Will squash after individual validations are reviewed

liggitt commented:
/cc @msau42 @jsafrane 
for storage reviews

/cc @tallclair @derekwaynecarr 
for node reviews

jsafrane commented:
storage validation lgtm

tallclair commented on pkg/apis/core/validation/ratcheting.go:
```diff
@@ -0,0 +1,124 @@
+		allErrs := field.ErrorList{}
+		visitContainers(data.(*core.PodSpec), fldPath, func(c *core.Container, fldPath *field.Path) {
+			for k, q := range c.Resources.Limits {
+				if k == core.ResourceMemory && q.MilliValue()%1000 != 0 {
```
Comment on `MilliValue` notes:
> // MilliValue returns the value of ceil(q * 1000); this could overflow an int64;
// if that's a concern, call Value() first to verify the number is small enough.

Can you just do: `_, loss := q.AsScale(0); loss` ?

liggitt replied:
done (though the loss bool was flipped)

tallclair commented:
/cc @dashpole 
/approve
for resource validation

nikhita commented:
/retest

liggitt commented:
I think all aspects have been reviewed/approved:
* Mechanism has lgtm from @sttts and @lavalamp 
* Storage has lgtm from @jsafrane 
* Resource has approval from @tallclair, can you lgtm as well?

tallclair commented:
/lgtm

bgrant0607 commented:
Sorry for not noticing this earlier.

I commented on #64841

liggitt commented:
/hold

bgrant0607 commented:
@liggitt Is the idea that every release the ratcheting validations would be moved to standard ones, and potentially new ratcheting validations would be introduced?

liggitt commented:
> Is the idea that every release the ratcheting validations would be moved to standard ones, and potentially new ratcheting validations would be introduced?

No, they couldn't be moved to be standard validations until we could guarantee the API servers would never encounter persisted etcd data that would fail those validations.

tallclair commented:
/lgtm cancel
See https://github.com/kubernetes/kubernetes/issues/64841#issuecomment-405441200

erictune commented:
Hi, sorry to be late to the party, but some here is some feedback on this approach:

- Yes, we should have CRD support for ratcheting, and it should ideally be very similar to the built-in support.
-  Annotating or modifying the CRD schema to express RatchetingValidations seems to me like it scales poorly.
- Instead of having special RatchetingValidation functions, instead, one could instead introduce a new api version with stronger validation.  Then one can set the desired storage version (as CRDs do) to lazily move over objects or run a "storage version upgrade script"  procedure to eagerly move over objects.
- To address the complaint that we don't want to move pods to v2, we could introduce a "microversion" and the validation uses both `apiversion` and `metadata.microversion` to compose a tuple which is used to select what functions/schema is used for validation.

liggitt commented:
> Instead of having special RatchetingValidation functions, instead, one could instead introduce a new api version with stronger validation. Then one can set the desired storage version (as CRDs do) to lazily move over objects or run a "storage version upgrade script" procedure to eagerly move over objects.

What would the storage migration do with existing objects that would not pass the new stricter validation?

As long as the previous API version is still usable (or data persisted via the previous API version is still able to be encountered), we have to handle round-tripping and updating that object via the new API version. Making an update of an unrelated field (like an annotation or a finalizer) fail for validation reasons is problematic.

bgrant0607 commented:
Quick note: We don't have versioned validation yet: #8550

mariantalla commented:
@hoegaarden and I were also wondering what you thought about the following:

1) The capability of also applying preprogrammed "migrations" to invalid objects? In other words, the next step after a failed validation could be to apply some logic to automatically reconcile it/ make it compatible with the new api version.

2) The validators being extracted from the body of the api server. This is perhaps more of a design consideration, but could make it easier to 
- maintain validator collections in future versions of the API server
- maybe even allow operators to define their own additional validators and configure the API server with those. Maybe admission webhooks already offer the tools for that use case though. 🤔

dims commented:
@liggitt should we shoot for v1.12?

liggitt commented:
> should we shoot for v1.12?

This isn't a top priority for the next week, so no

smarterclayton commented:
Fate of this?

lavalamp commented:
@smarterclayton It's solvable but with much more difficulty-- read from [this comment](https://github.com/kubernetes/kubernetes/issues/64841#issuecomment-413035093) to the end of the issue.

liggitt commented:
> @smarterclayton It's solvable but with much more difficulty-- read from [this comment](https://github.com/kubernetes/kubernetes/issues/64841#issuecomment-413035093) to the end of the issue.

If we limit ourselves to ratcheting in cases where the created objects were fatally flawed, I don't think we need to continue to accept invalid objects. I mostly closed this because I wasn't actively working on it.
